packages/relay-compiler/bin/RelayCompilerMain.js

```diff
@@ -97,163 +116,183 @@ type LanguagePlugin = PluginInitializer | {default: PluginInitializer};
+function getLanguagePlugin(
+  language: string | PluginInitializer,
+): PluginInterface {
   if (language === 'javascript') {
     return RelayLanguagePluginJavaScript();
   } else {
-    const pluginPath = path.resolve(process.cwd(), language);
-    const requirePath = fs.existsSync(pluginPath)
-      ? pluginPath
-      : `relay-compiler-language-${language}`;
-    try {
-      // eslint-disable-next-line no-eval
-      let languagePlugin: LanguagePlugin = eval('require')(requirePath);
-      if (languagePlugin.default) {
-        languagePlugin = languagePlugin.default;
-      }
-      if (typeof languagePlugin === 'function') {
-        return languagePlugin();
-      } else {
-        throw new Error('Expected plugin to export a function.');
+    let languagePlugin: LanguagePlugin;
+    if (typeof language === 'string') {
+      const pluginPath = path.resolve(process.cwd(), language);
+      const requirePath = fs.existsSync(pluginPath)
+        ? pluginPath
+        : `relay-compiler-language-${language}`;
+      try {
+        // eslint-disable-next-line no-eval
+        languagePlugin = eval('require')(requirePath);
+        if (languagePlugin.default) {
+          languagePlugin = languagePlugin.default;
+        }
+      } catch (err) {
+        const e = new Error(
+          `Unable to load language plugin ${requirePath}: ${err.message}`,
+        );
+        e.stack = err.stack;
+        throw e;
       }
-    } catch (err) {
-      const e = new Error(
-        `Unable to load language plugin ${requirePath}: ${err.message}`,
-      );
-      e.stack = err.stack;
-      throw e;
+    } else {
+      languagePlugin = language;
+    }
+    if (languagePlugin.default) {
+      languagePlugin = languagePlugin.default;
+    }
+    if (typeof languagePlugin === 'function') {
+      return languagePlugin();
+    } else {
+      throw new Error('Expected plugin to be a initializer function.');
     }
   }
 }
 
-async function main(options: {
-  schema: string,
-  src: string,
-  extensions: Array<string>,
-  include: Array<string>,
-  exclude: Array<string>,
-  verbose: boolean,
-  watchman: boolean,
-  watch?: ?boolean,
-  validate: boolean,
-  quiet: boolean,
-  persistOutput: ?string,
-  noFutureProofEnums: boolean,
-  language: string,
-  artifactDirectory: ?string,
-}) {
-  const schemaPath = path.resolve(process.cwd(), options.schema);
-  if (!fs.existsSync(schemaPath)) {
-    throw new Error(`--schema path does not exist: ${schemaPath}`);
+async function main(config: Config) {
+  if (config.verbose && config.quiet) {
+    throw new Error("I can't be quiet and verbose at the same time");
   }
-  const srcDir = path.resolve(process.cwd(), options.src);
-  if (!fs.existsSync(srcDir)) {
-    throw new Error(`--src path does not exist: ${srcDir}`);
+
+  config = getPathBasedConfig(config);
+  config = await getWatchConfig(config);
+
+  const codegenRunner = module.exports.getCodegenRunner(config);
```

Comment:
Other than adding optional function parameters to these functions to allow for DI, this appears to be _a_ suggestion to being able to mock functions that reside in the same module as the caller https://github.com/facebook/jest/issues/936#issuecomment-214939935.

I have no real preference for one or the other, let me know if you do.